// @mui/x-charts/ChartsRadialAxisHighlight/ChartsRadiusAxisHighlight.mjs
'use client';

import * as React from 'react';
import { useStore } from "../internals/store/useStore.mjs";
import { selectorChartPolarCenter, selectorChartRadiusAxis, selectorChartRotationAxis } from "../internals/plugins/featurePlugins/useChartPolarAxis/index.mjs";
import { selectorChartsInteractionPointerX, selectorChartsInteractionPointerY } from "../internals/plugins/featurePlugins/useChartInteraction/index.mjs";
import { ChartsRadialAxisHighlightCircle, ChartsRadialAxisHighlightPath } from "./ChartsRadialAxisHighlightPath.mjs";
import { jsx as _jsx } from "react/jsx-runtime";
function polarToSvg(cx, cy, radius, angle) {
  return [cx + radius * Math.sin(angle), cy - radius * Math.cos(angle)];
}

/**
 * @ignore - internal component.
 */
export default function ChartsRadiusAxisHighlight(props) {
  const {
    type,
    classes
  } = props;
  const store = useStore();
  const {
    cx,
    cy
  } = store.use(selectorChartPolarCenter);
  const {
    axis: radiusAxes,
    axisIds: radiusAxisIds
  } = store.use(selectorChartRadiusAxis);
  const {
    axis: rotationAxes,
    axisIds: rotationAxisIds
  } = store.use(selectorChartRotationAxis);
  const pointerX = store.use(selectorChartsInteractionPointerX);
  const pointerY = store.use(selectorChartsInteractionPointerY);
  if (pointerX === null || pointerY === null) {
    return null;
  }
  const radiusAxisId = radiusAxisIds[0];
  if (radiusAxisId === undefined) {
    return null;
  }
  const radiusAxis = radiusAxes[radiusAxisId];
  const radiusScale = radiusAxis.scale;
  const innerRadius = radiusScale.range()[0];
  const outerRadius = radiusScale.range()[1];
  const pointerRadius = Math.sqrt((pointerX - cx) ** 2 + (pointerY - cy) ** 2);
  if (pointerRadius < innerRadius || pointerRadius > outerRadius) {
    return null;
  }
  if (type === 'line') {
    const rotationAxisId = rotationAxisIds[0];
    const rotationAxis = rotationAxisId !== undefined ? rotationAxes[rotationAxisId] : undefined;
    const startAngle = rotationAxis?.scale.range()[0] ?? 0;
    const endAngle = rotationAxis?.scale.range()[1] ?? 2 * Math.PI;
    const isFullCircle = rotationAxis?.isFullCircle ?? Math.abs(endAngle - startAngle) >= 2 * Math.PI;
    if (isFullCircle) {
      return /*#__PURE__*/_jsx(ChartsRadialAxisHighlightCircle, {
        cx: cx,
        cy: cy,
        r: pointerRadius,
        className: classes.root,
        ownerState: {
          axisHighlight: 'line'
        }
      });
    }
    const [startX, startY] = polarToSvg(cx, cy, pointerRadius, startAngle);
    const [endX, endY] = polarToSvg(cx, cy, pointerRadius, endAngle);
    const isLargeArc = Math.abs(endAngle - startAngle) >= Math.PI;
    const isDirectArc = endAngle > startAngle;
    return /*#__PURE__*/_jsx(ChartsRadialAxisHighlightPath, {
      d: `M ${startX} ${startY} A ${pointerRadius} ${pointerRadius} 0 ${isLargeArc ? 1 : 0} ${isDirectArc ? 1 : 0} ${endX} ${endY}`,
      className: classes.root,
      ownerState: {
        axisHighlight: 'line'
      }
    });
  }
  return null;
}
